[Key Vault]Change keyvault to track2

zedy-wj opened the pull request:
**Project ClearClipboard and Core**

Only change TargetFrameworkVersion from 4.5.1 to 4.6.1

**Project Library**

KeyVaultClient splits into SecretClient and CertificateClient
Update track1 secret api from track1 to track2
Update track1 certificate api track1 to track2
Change TargetFrameworkVersion from 4.5.1 to 4.6.1
Migrate to Azure.Identity from ASAL

**Project Explorer**

Change track1 secret model from track1 to track2.
Change track1 certificate model from track1 to track2.
Change using track1 Project Library api to track2.
Some track1 key vault model splits into secret model and certificate model.
Change TargetFrameworkVersion from 4.5.1 to 4.6.1
Migrate to Azure.Identity from ASAL

jongio commented:
@heaths - We have KV/.NET code here.  Would be great if you could review.
@christothes - We have some identity code updates here. Would be great if you could review.

Thanks!

heaths commented on Vault/Core/Core.csproj:
```diff
@@ -15,6 +15,8 @@
     <RootNamespace>Microsoft.Vault.Core</RootNamespace>
     <AssemblyName>Microsoft.Vault.Core</AssemblyName>
     <FileAlignment>512</FileAlignment>
+    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
```
Given they're dependent on WinForms, targeting the least common denominator for .NET Framework is advisable. It'll run on newer .NET Frameworks which, for v4, are a global install.

zedy-wj commented on Vault/Explorer/CertificateDialog.cs:
```diff
@@ -73,40 +72,31 @@ public CertificateDialog(ISession session, X509Certificate2 cert) : this(session
             uxTextBoxName.ReadOnly = true;
             int i = 0;
-            uxMenuVersions.Items.AddRange((from v in versions orderby v.Attributes.Created descending select new CertificateVersion(i++, v)).ToArray());
+            uxMenuVersions.Items.AddRange((from v in versions orderby v.CreatedOn descending select new CertificateVersion(i++, v)).ToArray());
```
Not very clear about the code logic here, we just changed the keyvault. And I haven't found an `AddRange()` method in the `ToolStripItemCollection` class that can receive the `IEnumerable<T>`. If we need to change it, can you be more detailed? Thanks!

zedy-wj commented on Vault/Library/CachePersistence.cs:
```diff
@@ -0,0 +1,133 @@
+
+namespace Microsoft.Vault.Library
+{
+    public class CachePersistence
```
I referenced this document and made modifications, please see [here](https://github.com/zedy-wj/AzureKeyVaultExplorer/blob/zedy-keyvault/Vault/Library/CachePersistence.cs#L93-L131) for details. What confuses you about this modification?

zedy-wj commented:
@christothes - Changes to the token cache involve other changes, so we'll update that first. I have replied your comments in this PR, looking forward to your review, thanks.

christothes commented on .gitignore:
```diff
@@ -11,3 +11,4 @@ msbuild.log
 *.tmp
 ~*.docx
 ~*.xlsx
+packages/*
```
This doesn't seem to be configured correctly, since the PR includes all the assemblies in your nuget cache. 
```suggestion
packages
```

zedy-wj replied:
Thanks for your suggestion, all the current assemblies were added when .gitignore was not configured before, we will remove all the assemblies in the next update. and we'll resolve heaths's comments ASAP.

christothes reviewed (commented):
Please address @heaths 's comments as well and then I will take another look.

zedy-wj commented on Vault/Library/CachePersistence.cs:
```diff
@@ -0,0 +1,133 @@
+
+                // Construct a new client with our TokenCachePersistenceOptions with the addition of the AuthenticationRecord property.
+                // This tells the credential to use the same token cache in addition to which account to try and fetch from cache when GetToken is called.
+                var credential = new InteractiveBrowserCredential(
```
The modification here we refered to [document](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/samples/ClientSideUserAuthentication.md#persisting-the-authenticationrecord), and browser credential was used in the sample. Also, we didn't find a way to use TokenCachePersistenceOptions in `DefaultAzureCredential`.

zedy-wj commented on Vault/Library/CachePersistence.cs:
```diff
@@ -0,0 +1,133 @@
+            string[] tokenNames = GetAllFileTokenCacheLoginNames();
+            foreach (string token in tokenNames)
+            {
+                new CachePersistence(token).Clear();
```
Direct deletion is a call to a non-static method that requires a heap allocation. The `ClearAllFileTokenCaches `method is also called in other non-static methods, and we don't want to change its type here.

zedy-wj commented:
@christothes - We reverted all assemblies and updated codes according to your comments. Several of the remaining comments are related to the identity token cache. Regarding the use of `DefaultAzureCredential `instead of `InteractiveBrowserCredential`, we are not sure how to use `TokenCachePersistenceOptions` in `DefaultAzureCredential`, is it now supported?

v-xuto commented:
@christothes  We have updated by your comments. Please re-review this PR. Thanks a lot.

zedy-wj commented:
@heaths - Regarding the part of the comments that are not resolved, we are upgrading the `TokenCache` from ADAL to Azure.Identity. So far, we've only seen ways to use `TokenCachePersistenceOptions ` in `InteractiveBrowserCredential`, and haven't found how to use it in `DefaultAzureCredential`, if you're clear about this, could you give a sample, thanks in advance!

@christothes - As mentioned above, not sure whether `TokenCachePersistenceOptions ` is supported in `DefaultAzureCredential`. If so, how to use it?

christothes commented:
> @christothes - As mentioned above, not sure whether `TokenCachePersistenceOptions ` is supported in `DefaultAzureCredential`. If so, how to use it?

We recently realized that we cannot use `TokenCachePersistenceOptions` with `DefaultAzureCredential`. To work around this for the sample, we can use `ChainedTokenCredential`.

zedy-wj commented:
@christothes - Do you mean to initialize both `InteractiveBrowserCredential` and `DefaultAzureCredential` as parameters to a `ChainedTokenCredential`?

Such as:
```
                var credential = new InteractiveBrowserCredential(
                    new InteractiveBrowserCredentialOptions
                    {
                        TokenCachePersistenceOptions = new TokenCachePersistenceOptions { Name = TOKEN_CACHE_NAME },
                        AuthenticationRecord = authRecord
                    });

                var chainedTokenCredential = new ChainedTokenCredential(credential, new DefaultAzureCredential());
```

zedy-wj commented:
@christothes -  So how do we connect `ChainedTokenCredential` to `DefaultAzureCredential`? We are not familiar with `ChainedTokenCredential`, is there any relevant documentation?

christothes commented:
> @christothes - After investigating, the `ChainedTokenCredential` class provides the ability to link together multiple credential instances to be tried sequentially when authenticating. We have the following questions:
> 
> 1. Actually, we cannot use `TokenCachePersistenceOptions` with `DefaultAzureCredential`, so there is only `InteractiveBrowserCredential` we can pass. In this case, what the meaning of using `ChainedTokenCredential` in this sample? Why don't we use `InteractiveBrowserCredential` directly?
> 2. Our current understanding of using `ChainedTokenCredential` here is that it will be easier to expand after the `TokenCachePersistenceOptions` is supported in DAC. (If we omit or misinterpret your meaning, please add)
> 
> Notes: The idea of suggesting a DAC here was originally proposed by heaths, we're just asking here if it's feasible.

If the intent of this example is to prompt the user interactively for credentials and then cache them for silent authentication, then all you need is `InteractiveBrowserCredential` as shown in this example: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/samples/ClientSideUserAuthentication.md#silent-authentication-with-authenticationrecord-and-tokencachepersistenceoptions

zedy-wj commented:
> > @christothes - As mentioned above, not sure whether `TokenCachePersistenceOptions ` is supported in `DefaultAzureCredential`. If so, how to use it?
> 
> We recently realized that we cannot use `TokenCachePersistenceOptions` with `DefaultAzureCredential`. To work around this for the sample, we can use `ChainedTokenCredential`.

@heaths - The `DefaultAzureCredential` doesn't support our features here, so we haven't modified it based on your comments, at the moment we're only seeing `InteractiveBrowserCredential` support `TokenCachePersistenceOptions`, do you have any ideas?

heaths commented:
Most of the other credentials used by `DefaultAzureCredential` don't need to cache: most are already persistent e.g., VS and VSCode credentials, Azure CLI and Az PowerShell credentials, etc. If people are already authenticated with those, there's no need to prompt again.

Doesn't seem like `DefaultAzureCredentialOptions` allows you to set individual `InteractiveBrowserCredentialOptions` - or any other credential options - though. So you could either used a `ChainedTokenCredential` with a [similar order as DefaultAzureCredential](https://docs.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet#definition), or continue supporting just the `InteractiveBrowserCredential`.

zedy-wj commented:
@heaths - For now we will continue supporting just the `InteractiveBrowserCredential`, and then we will use `ChainedTokenCredential` when `TokenCachePersistenceOptions` are supported in other criedentials.

We will not make any changes to the two comments you left about `DefaultAzureCredential`, and if there are no other questions, we will click on the `resolve conversation`.

zedy-wj commented:
@heaths - We have updated according to your comments except two you left about `DefaultAzureCredential`. Please re-review it, thanks!

heaths commented on Vault/Library/CachePersistence.cs:
```diff
@@ -124,10 +128,93 @@ void AfterAccessNotification(AuthenticationRecord authRecord)
+
+            // Create a new instance of the RNGCryptoServiceProvider.
+            // Fill the array with a random value.
+            new RNGCryptoServiceProvider().GetBytes(entropy);
```
This is disposable so you should use something like `using var rng = new RNGCryptoServicePRovider();` You could also save some time by declaring this as a `static readonly` field instead of initializing it with every call.

heaths commented on Vault/Library/CachePersistence.cs:
```diff
@@ -124,10 +128,93 @@ void AfterAccessNotification(AuthenticationRecord authRecord)
                 authRecord = credential.Authenticate();
 
+                // Create the original data to be encrypted
+                var toEncrypt = Encoding.ASCII.GetBytes(authRecord.ToString());
```
Could lead to data corruption and transcoding issues. Should use `Encoding.UTF8`.

heaths commented on Vault/Library/CachePersistence.cs:
```diff
@@ -124,10 +128,93 @@ void AfterAccessNotification(AuthenticationRecord authRecord)
+            return length;
+        }
+
+        public static byte[] DecryptDataFromStream(byte[] Entropy, DataProtectionScope Scope, Stream S, int Length)
```
I don't see where this is getting called.

zedy-wj replied:
When decrypting data, we need an `AuthenticationRecord` parameter to implement silent authentication.
```
// Load the previously serialized AuthenticationRecord from disk and deserialize it.
var authRecordStream = new FileStream(FileName, FileMode.Open, FileAccess.Read);
authRecord = AuthenticationRecord.Deserialize(authRecordStream);

// Construct a new client with our TokenCachePersistenceOptions with the addition of the AuthenticationRecord property.
// This tells the credential to use the same token cache in addition to which account to try and fetch from cache when GetToken is called.
var credential = new InteractiveBrowserCredential(
    new InteractiveBrowserCredentialOptions
    {
        TokenCachePersistenceOptions = new TokenCachePersistenceOptions { Name = TOKEN_CACHE_NAME },
        AuthenticationRecord = authRecord
    });
```

But `ProtectedData.Unprotect(inBuffer, Entropy, Scope);` method return an `byte[] array` value. So we didn't call here, considering the possible updates after that, we kept it for the time being, do you have any ideas or suggestions?

heaths replied:
If you're not decrypting your cache, I don't see how it's useful. You could instead initially a `MemoryStream` from a `byte[]` array which will give you the stream you need.

zedy-wj commented:
@heaths - Updated according to your comments, please re-review this PR, thanks!

heaths commented on Vault/Library/CachePersistence.cs:
```diff
@@ -89,17 +91,15 @@ public void Rename(string newName, AuthenticationRecord authRecord)
                 var authRecordStream = new FileStream(FileName, FileMode.Open, FileAccess.Read);
-                authRecord = AuthenticationRecord.Deserialize(authRecordStream);
+                byte[] decryptData = DecryptDataFromStream(entropy, DataProtectionScope.CurrentUser, authRecordStream, bytesWritten);
+                MemoryStream memoryStream = new MemoryStream(decryptData);
```
```suggestion
                using MemoryStream memoryStream = new MemoryStream(decryptData);

```

You should always dispose disposables when finished.

zedy-wj replied:
Updated according to your comments.
Thanks for your guidance :)

heaths reviewed (approved):
@christothes this look good to you? Seems to follow the example pattern.

zedy-wj commented:
@christothes - We have done an update on the comments of heaths, could you re-review this PR, thanks!

v-xuto commented:
@heaths We have updated by your comments, please re-review this PR.

v-xuto commented:
@christothes Could you please re-review this PR? If you have no questions about this PR, please approve.

v-xuto commented:
@jongio Do you have permission to merge this PR? If so, please help merge.

jongio commented:
I don't, but I'm working on making contact with the team.

v-xuto commented:
@jongio Regarding merge this PR, is there any progress so far?

v-xuto commented:
@kevin-msft Do you have permission to merge this PR? If so, please help merge.

v-xuto commented:
@jongio  Help find the corresponding person merge this PR. Thanks in advance.
